core/src/main/java/org/elasticsearch/index/shard/StoreRecovery.java

```diff
@@ -393,10 +398,9 @@ private void internalRecoverFromStore(IndexShard indexShard) throws IndexShardRe
                     logger.debug("failed to list file details", e);
                 }
                 indexShard.performTranslogRecovery(indexShouldExists);
-                assert indexShard.shardRouting.primary() : "only primary shards can recover from store";
                 indexShard.getEngine().fillSeqNoGaps(indexShard.getPrimaryTerm());
+                indexShard.finalizeRecovery(indexShard.getLocalCheckpoint());
```

Comment:
This looks unsafe to me.

Assume a situation where you have a primary with an in-sync replica. Now both nodes go down while the primary is replicating writes to the replica, i.e. global checkpoint is below local checkpoint on primary.

The node with the original primary comes up, and moves the global checkpoint to its local checkpoint. No writes are happening on the primary shard, so the replica stays as in-sync. The node with the primary then goes down. The node with the replica comes up and gets assigned as primary (because it remained in-sync). It then has the possibility to write different docs into slots that are below the previous primary's global checkpoint.